### PR TITLE
Esql/create DATE_DIFF function

### LINKED ISSUE #101942: ESQL: add DATE_DIFF functionality
### Description

Similar to what [ES SQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-diff) has, we need functionality for datetime related data types difference: `eval duration=@timestamp-timestamp_end` or `eval duration=DATE_DIFF("year", @timestamp, timestamp_end)`.

### PR DESCRIPTION
**Pull request by nicolasgras**
This pull request resolves https://github.com/elastic/elasticsearch/issues/101942.

A new functionality DATE_DIFF is added to [ESQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql.html). This functionaliy is similar to the ones in [SQL](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-diff) plugin.

It calculates the diffenrence between this 2 dates using a unit specified and takes 3 input parameters:
- unit (text)
- startTimestamp (dateTime)
- endTimestamp (dateTime)

Screenshot from 2023-12-08 01-37-41

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    @Evaluator
+    static int process(BytesRef unit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        ZonedDateTime zdtStart = ZonedDateTime.ofInstant(Instant.ofEpochMilli(startTimestamp), UTC);
+        ZonedDateTime zdtEnd = ZonedDateTime.ofInstant(Instant.ofEpochMilli(endTimestamp), UTC);
+
+        Part datePartField = Part.resolve(unit.utf8ToString());
+        if (datePartField == null) {
+            List<String> similar = Part.findSimilar(unit.utf8ToString());
+            if (similar.isEmpty()) {
+                throw new InvalidArgumentException(
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java:144]: A nit on style: you could prepare the message in the two branches and throw just once.

**Review by bpintea — COMMENTED** [review]
Thanks for your interest and contribution, @nicolasgras.
I had a first pass. My main question is about the code duplication. Any reason you didn't extract the common parts (in the ql module) and reused it here?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        public Integer diff(ZonedDateTime startTimestamp, ZonedDateTime endTimestamp) {
+            return diffFunction.apply(startTimestamp, endTimestamp);
+        }
+
+        @Override
+        public Iterable<String> aliases() {
+            return aliases;
+        }
+
+        private static int safeInt(long diff) {
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I have updated the code to use DataTypeConverter#safeToInt and removed the one I implemented that was identical.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        MILLISECOND((start, end) -> safeInt(ChronoUnit.MILLIS.between(start, end)), "milliseconds", "ms"),
+        MICROSECOND((start, end) -> safeInt(ChronoUnit.MICROS.between(start, end)), "microseconds", "mcs"),
+        NANOSECOND((start, end) -> safeInt(ChronoUnit.NANOS.between(start, end)), "nanoseconds", "ns");
+
+        private static final Map<String, Part> NAME_TO_PART;
+        private static final List<String> VALID_VALUES;
+
+        static {
+            NAME_TO_PART = DateTimeField.initializeResolutionMap(values());
+            VALID_VALUES = DateTimeField.initializeValidValues(values());
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I agree, VALID_VALUES List is not used. I should have cleant this part.
The initialization of NAME_TO_PART is inlined now.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        private BiFunction<ZonedDateTime, ZonedDateTime, Integer> diffFunction;
+        private Set<String> aliases;
+
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I agree, they are final now.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        public static Part resolve(String dateTimeUnit) {
+            return DateTimeField.resolveMatch(NAME_TO_PART, dateTimeUnit);
+        }
+    }
+
+    @FunctionInfo(returnType = "integer",
+        description = "Subtract 2 dates and return their difference in multiples of a unit specified in the 1st argument")
+    public DateDiff(
+        Source source,
+        @Param(name = "unit", type = { "text" }, description = "A valid date unit") Expression unit,
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I agree, I have updated the code.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        ) Expression startTimestamp,
+        @Param(name = "endTimestamp", type = { "date" }, description = "A string representing a end timestamp") Expression endTimestamp
+    ) {
+        super(source, List.of(unit, startTimestamp, endTimestamp));
+        this.unit = unit;
+        this.startTimestamp = startTimestamp;
+        this.endTimestamp = endTimestamp;
+    }
+
+    @Evaluator
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I agree, I have updated the code.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    static int process(BytesRef unit, long startTimestamp, long endTimestamp) throws IllegalArgumentException {
+        ZonedDateTime zdtStart = ZonedDateTime.ofInstant(Instant.ofEpochMilli(startTimestamp), UTC);
+        ZonedDateTime zdtEnd = ZonedDateTime.ofInstant(Instant.ofEpochMilli(endTimestamp), UTC);
+
+        Part datePartField = Part.resolve(unit.utf8ToString());
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I have updated the code by adding a new evaluator (I get something similar to DateParse now).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        description = "Subtract 2 dates and return their difference in multiples of a unit specified in the 1st argument")
+    public DateDiff(
+        Source source,
+        @Param(name = "unit", type = { "text" }, description = "A valid date unit") Expression unit,
+        @Param(
+            name = "startTimestamp",
+            type = { "date" },
+            description = "A string representing a start timestamp"
+        ) Expression startTimestamp,
+        @Param(name = "endTimestamp", type = { "date" }, description = "A string representing a end timestamp") Expression endTimestamp
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: This typo has been fixed.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java (hunk 1 above)
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: I agree, I have updated the code.

**Comment by nicolasgras** [comment]
> Thanks for your interest and contribution, @nicolasgras. I had a first pass. My main question is about the code duplication. Any reason you didn't extract the common parts (in the ql module) and reused it here?

Thank you for you review @bpintea.
As this issue is about **esql**, I thought I should not update any other plugin. But it is better to use it in a common module.
As you suggest, I have moved **DateTimeFiled** interface in **ql** module. Maybe let's keep in mind to also use it in sql module if this PR is accepted and merged.

I have also added unit tests in esql/qa/testFixtures directory to have a better code coverage.

Let me know if it is ok for you.

**Review by luigidellaquila — COMMENTED** [review]
Thank you very much @nicolasgras!
Just two comments from my side:
- I see the CI still complains about CLA signature; I checked your commits and I see they appear with two different email addresses (one is invalid of course):

```
Author: Nicolas Gras <nicolasgras01@gmail.com>
Author: Nicolas Gras <nicolas@system76-pc.localdomain>
```

You have to sign the CLA with both to make the checks green, even though the second one is obviously a local one.

- I think @bpintea referred to the content of `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff` (the `enum Part` and some logic) when he asked about common parts that can be extracted to QL.
IMHO it's not mandatory, we could also keep the two things separated and have this PR self-contained in ESQL module, but in that case I'd avoid to move DateTimeField to QL

**Comment by nicolasgras** [comment]
Thank you @luigidellaquila for your review.

> You have to sign the CLA with both to make the checks green, even though the second one is obviously a local one.  

I have rebased my commits on top of the branch and updated the author of them, then the CI does not complain anymore about CLA signature.

>  I think @bpintea referred to the content of `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff` (the `enum Part` and some logic) when he asked about common parts that can be extracted to QL.
>       IMHO it's not mandatory, we could also keep the two things separated and have this PR self-contained in ESQL module, but in that case I'd avoid to move DateTimeField to QL.

@bpintea , what do you think about it?
After your review I updated and cleant `org.elasticsearch.xpack.esql.expression.function.scalar.date.DateDiff.Part`, even if it is similar to `org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateDiff.Part`, they are a little different now.
And do you think we should move back `DateTimeField` to ESQL?

**Comment by bpintea** [comment]
> @bpintea , what do you think about it?

In some cases sharing code between other *ql modules within the `ql` one makes things restrictive/tricky. If the code is very similar, it makes sense, however. But I see you've used some SDK functionality over custom implementation, which is preferred, and if that covers the needs for this PR, I'm OK containing the changes to the `esql` module, as @luigidellaquila suggested. (We can extract these changes in QL and update SQL later if needed.)

**Comment by nicolasgras** [comment]
@bpintea and @luigidellaquila, I have moved back `DateTimeField` interface into **ESQL**.
An issue could be created later to extract some code in **QL** and rework some parts in **SQL** if needed.
Is it ok for you?

**Review by bpintea — COMMENTED** [review]
I've left two notes, otherwise it LGTM.

I guess with SQL's DateDiff we've follow other implementations returning int and we follow suite also here. This is OK, though we might want to update it, if we'll add support for nano timestamps (which we don't have yet), in which case an int will not cover even 3s.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            if (datePartField == null) {
+                List<String> similar = DateTimeField.findSimilar(NAME_TO_PART.keySet(), dateTimeUnit);
+                String errorMessage = String.format(
+                    Locale.ROOT,
+                    "A value of %s or their aliases is required; received [%s]",
+                    Arrays.asList(Part.values()),
+                    dateTimeUnit
+                );
+                if (similar.isEmpty() == false) {
+                    errorMessage = String.format(
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: ok, I have updated the code

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        ExpressionEvaluator.Factory startTimestampEvaluator = toEvaluator.apply(startTimestamp);
+        ExpressionEvaluator.Factory endTimestampEvaluator = toEvaluator.apply(endTimestamp);
+
+        if (unit.foldable()) {
+            try {
+                Part datePartField = Part.resolve(((BytesRef) unit.fold()).utf8ToString());
+                return new DateDiffConstantEvaluator.Factory(source(), datePartField, startTimestampEvaluator, endTimestampEvaluator);
+            } catch (IllegalArgumentException e) {
+                throw new EsqlIllegalArgumentException(e, "invalid unit format for [{}]: {}", sourceText(), e.getMessage());
```
[the remark this answers is not in the document]
Comment (nicolasgras) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateDiff.java]: ok, I have updated the type of Exception thrown here to `InvalidArgumentException`

**Review by bpintea — APPROVED** [review]
Thanks, LGTM.

**Review by luigidellaquila — APPROVED** [review]
Thank you very much @nicolasgras, LGTM

**Comment by DaveCTurner** [comment]
Unfortunately this commit broke the build (checkstyle violations) so I reverted it in 943b2eae70e7395ddd6b4a8c4ef36972082ab018. Luigi will open a PR to reinstate it.

**Comment by luigidellaquila** [comment]
https://github.com/elastic/elasticsearch/pull/104118

**Pull request merged**